src/utils/gatekeeper.hpp

```diff
@@ -159,55 +179,62 @@ struct Gatekeeper {
     }
 
     // Completely invalidated the accessor if return true
-    [[nodiscard]] bool try_delete(std::chrono::milliseconds timeout = std::chrono::milliseconds(100)) {
+    template <typename Func = decltype([](T &) { return true; })>
+    [[nodiscard]] bool try_delete(std::chrono::milliseconds timeout = std::chrono::milliseconds(100),
+                                  Func &&predicate = {}) {
       // Prevent new access
       auto guard = std::unique_lock{owner_->mutex_};
       if (!owner_->cv_.wait_for(guard, timeout, [this] { return owner_->count_ == 1; })) {
         return false;
       }
-      // Delete value
+      // Already deleted
+      if (owner_->value_ == std::nullopt) return true;
+      // Delete value if ok
+      if (!predicate(*owner_->value_)) return false;
       owner_->value_ = std::nullopt;
       return true;
     }
 
-    explicit operator bool() const { return owner_ != nullptr; }
+    explicit operator bool() const {
+      return owner_ != nullptr         // we have access
+             && !owner_->is_deleting;  // AND we are allowed to use it
+    }
 
     void reset() {
       if (owner_) {
         {
           auto guard = std::unique_lock{owner_->mutex_};
           --owner_->count_;
         }
-        owner_->cv_.notify_all();
+        owner_->cv_.notify_all();  // TODO Under a lock???
```

Comment:
comment not needed
> The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s). Doing so may be a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock, though some implementations recognize the pattern and do not attempt to wake up the thread that is notified under lock.